README.md
# Keysmith

Create keys, uuid's, timestamps, and more with Rust.

Here's some example keys:

```text
5WdT@KfYRyRDqh0AAVUsyitSXbm4OQwzClX9.XYn5kLmKeObCf8YE3HqzHdU3UTS
g1Y_Zp9-9rDf7VAaR1KGnH9Uf5klGjSQ6P2rAjK35iK-GnZ@dVXYu.aElzTfTOaK
M3rkxIbrD0lz-SpUBe704X2)Gd(_WprzRJW7N51O@_58180Gs9esIVBO5.OdFOlY
```

This looks like gibberish, but that's the point.
Two id's in a database should not overlap.
So if you need a lot of entries, you need to reduce the odds of that happening.
If the odds of two identical id's are astronomically low,
you shouldn't even need to check for an existing id in the first place.
Keysmith does that generation for you and outputs it as a String.

You can also generate version 4 UUID's.

```text
be3b5529-931b-6b75-1678-a057bccf71c9
440146ab-a19a-8a36-2d4b-ba19a79570d4
0090da40-6ce2-59d3-629d-11b293c9e2d3
```

Keysmith also supports many other data formats.
- Keys strings from custom character sets
- "Nonstandard" V4 UUID's using any letter as opposed to just a-f
- Timestamps (ex: 50546851)
- Dates (ex: 2023-07-17), times (ex: 19:34:07), and datetimes (ex: 2023-07-17T19:34:07)
- Individual characters

## Usage

For usage, see the [documentation](https://docs.rs/keysmith/latest/keysmith/).

For installation, see [crates.io](https://crates.io/crates/keysmith).

## Changelog

You can find the changelog in [CHANGELOG.md](CHANGELOG.md).

## Building

1. Install [rustup](https://www.rust-lang.org/tools/install) if you haven't already.
2. Clone this repo. Usually with `git clone https://github.com/njshockey/keysmith-rs.git`.
3. Run `cargo build` or `cargo build --release` to build.
   - To run the test package, run `cargo run -p usage` instead.

## License

Keysmith uses the Rust standard MIT/Apache-2.0 dual license for best
compatibility. See
[LICENSE-APACHE.txt](LICENSE-APACHE.txt) and [LICENSE-MIT.txt](LICENSE-MIT.txt)
for the full licenses.


CHANGELOG.md
# Changelog

For the source code of each version, see the [tags page](https://github.com/njshockey/keysmith-rs/tags).

## v0.4.0 (August 23, 2023)
**_CAUTION_**: v0.4.0 introduces many breaking API changes. It's a change for the better as the API is much simpler now, but be careful before upgrading. This should be the last major API change.

API
- Publicly accessed functions have had their gen_ or get_ prefixes removed. This looks like gen_key() => key().

Features
- Generate datetime time stamps.
  - date::datetime() -> 2023-07-17T19:34:07
  - date::datetime_pretty() -> 2023/07/17 19:34:07
  - date::datetime_no_seps() -> 20230717193407
  - See date module for more.
- Generate timestamps using milliseconds.
  - timestamp::timestamp_ms() -> 50546851978
- Generate UUID's without separators.
  - uuid::uuid4_no_seps() -> fc402d5270be7f09caed8da65db08985
- Generate keys from custom character sets.
  - key::key_custom(length, charset)
  - char::char_custom(length, charset)

Development
- Private functions are not effected by new API. (Like key::gen_char_from_opts().)
- Chars are now generated using pub consts instead of a hashmap.
- Lots of code cleanup.
- Unit tests.


## v0.3.1 (January 29, 2022)

**_WARNING_**: v0.4.0 will have breaking API changes.
Functions will have their gen or get prefixes removed.
So gen_key() will become key().
This should make the API much easier to use but keep it in mind for existing projects.
v0.3.1 has not made this change yet.

Features

- Can now generate timestamps with custom epochs.
    - Use get_timestamp_custom(epoch: &str)
    - Epoch format = "Sat, 1 Jan 2022 00:00:00 +0000".
- Can now generate timestamps from the first second of 1970 (utc).
    - Use get_timestamp_utc()

Development

- Nicer formatting thanks to [Rustfmt](https://rust-lang.github.io/rustfmt/).

## v0.3.0 (October 28, 2022)

**_CAUTION_**: v0.3.0 introduces breaking changes. Mainly with importing modules.
Be careful when upgrading from an older version.

Features

- Can now get a timestamp for the current time.
    - Use get_timestamp() or get_timestamp_i64() for this.
    - Based on number of seconds since the first second of 2022.
- Can now generate keys with only numbers and letters.
    - Use gen_nums_and_letters() for this.
    - gen_nums_and_letters_lower() for just lowercase letters.
    - gen_nums_and_letters_upper() for just uppercase letters.
- Generate a single character with the gen_char module.
    - gen_char() to generate a character found in gen_key().
    - gen_uuid_char() to generate a character found in gen_uuid().

API

- Functions have been separated into different modules.
Import relevant features with keysmith::{key, timestamp, uuid, gen_char}
where necessary. See [docs](https://docs.rs/keysmith/latest/keysmith/).

Docs

- Rust generated auto docs. Link to newest version can be found on [docs.rs](https://docs.rs/keysmith/latest/keysmith/).

Testing

- Can now toggle individual modules and functions.
- Cleaned up ks-tests crate.

Development

- gen_char() and gen_uuid_char() are now public
and can be used with the keysmith crate.
- Cleaned up a lot of code under the hood.

## v0.2.1 (October 20, 2022)

You can now generate keys with only specific kinds of characters.
For more details, see [usage](README.md#usage).

- gen_nums() to generate only numbers.
- gen_letters() to generate only letters (either uppercase or lowercase).
- gen_letters_lower() to generate only lowercase letters.
- gen_letters_upper() to generate only uppercase letters.
- gen_special_chars() to generate only safe special characters.
- gen_special_chars_unsafe() to generate only unsafe special characters.
    - WARNING: These are unsafe because they tend to break
      file structures and URL's and the like.
    - This is not recommended but might be useful to _someone_.

## v0.2.0 (October 18, 2022)

- Can now generate version 4 UUID's with gen_uuid('4'). See [usage](README.md#usage).
- "Nonstandard" UUID's can also be generated with gen_uuid('n'). See [usage](README.md#usage).

## v0.1.1 (October 7, 2022)

- gen_key() can now generate uppercase letters.
- Fixed a build error on Linux.

## v0.1.0 (October 6, 2022)

- First release.
- Generate a key with gen_key(). See [usage](README.md#usage).


keysmith/src/lib.rs
//! Create keys, uuid's, timestamps, and more with Rust..
//!
//!
//! Here's some example keys:
//!
//! ```text
//! 5WdT@KfYRyRDqh0AAVUsyitSXbm4OQwzClX9.XYn5kLmKeObCf8YE3HqzHdU3UTS
//! g1Y_Zp9-9rDf7VAaR1KGnH9Uf5klGjSQ6P2rAjK35iK-GnZ@dVXYu.aElzTfTOaK
//! M3rkxIbrD0lz-SpUBe704X2)Gd(_WprzRJW7N51O@_58180Gs9esIVBO5.OdFOlY
//! ```
//!
//! This looks like gibberish, but that's the point. Two id's in a database should not overlap.
//! So if you need a lot of entries, you need to reduce the odds of that happening.
//! If the odds of two identical id's are astronomically low, you shouldn't even need to check for an existing id in the first place.
//! Keysmith does that generation for you and outputs it as a String.
//!
//! You can also generate version 4 UUID's. Examples:
//! ```text
//! be3b5529-931b-6b75-1678-a057bccf71c9
//! 440146ab-a19a-8a36-2d4b-ba19a79570d4
//! 0090da40-6ce2-59d3-629d-11b293c9e2d3
//! ```
//!
//! Keysmith also supports many other data formats.
//! - Keys strings from custom character sets
//! - "Nonstandard" V4 UUID's using any letter as opposed to just a-f
//! - Timestamps (ex: 50546851)
//! - Dates (ex: 2023-07-17), times (ex: 19:34:07), and datetimes (ex: 2023-07-17T19:34:07)
//! - Individual characters
//!
//! # Usage
//!
//! First, add this to your `Cargo.toml` dependencies.
//! Check out the [crate page](https://crates.io/crates/keysmith) to make sure you're using the latest version.
//!
//! ```toml
//! [dependencies]
//! keysmith = "0.4.0"
//! ```
//! Next, in the actual code:
//!
//! ```
//! use keysmith::{key, uuid};
//!
//! let key1: String = key::key(64);
//! println!("{}", key1);
//! // ex: aVCkmMFkv3UqLIU2tC52DQOWrIg4RsaH.xvTCTvA_PVHY@MOIWH~y1610KIZ@qM@
//!
//! let key2: String = uuid::uuid();
//! println!("{}", key2);
//! // ex: da748b8b-e915-661b-466a-0d6a0480112a
//!
//! let key3: String = uuid::uuidn();
//! println!("{}", key3);
//! // "Nonstandard" UUID
//! // ex: eko0c6ph-k2ok-60rr-pj78-mns182t9vurf
//! ```
//! The first function generates a key with a length of 64 characters as a String.
//! The second generates a version 4 UUID.
//! The third generates a "nonstandard" UUID.
//! It uses the same structure as a version 4, but it can have any letter instead of just a-f.
//! The "nonstandard" version is obviously not standard,
//! but may be useful depending on your preferences.
//!
//!
//! You can also generate a key consisting of only a specific kind of character.
//!
//! ```
//! use keysmith::key;
//!
//! let key1: String = key::key(32);
//! println!("{}", key1);
//! // ex: 04356417134317004828941212534445
//!
//! let key2: String = key::letters(32);
//! println!("{}", key2);
//! // ex: PBSZWwSTmRalGnzeSbQUKmXRikKUWXvj
//!
//! let key3: String = key::special_chars(32);
//! println!("{}", key3);
//! // Special characters generally considered "safe."
//! // Possible characters: -_.()~@
//! // ex: )@-_~@_@._))~)@))@.)(-)@(.@(~((@
//! ```
//!
//! If you want only lowercase or uppercase letters, you can do that, too!
//!
//! ```
//! use keysmith::key;
//!
//! let key1: String = key::letters_lower(32);
//! println!("{}", key1);
//! // ex: xoewhgvjsqzctfgpaqwnhanbgweflpqc
//!
//! let key2: String = key::letters_upper(32);
//! println!("{}", key2);
//! // ex: EVQMPIHKDBPLZJBPCHTXTIBLYRSFFFUY
//! ```
//!
//! If you need something more specific than what's already here by default,
//! you can generate a single char.
//!
//! ```
//! use keysmith::char::{char, GenCharOpts};
//!
//! let opts = GenCharOpts {
//!     nums: true,
//!     letters: true,
//!     upper: true,
//!     safe_sp_chars: true,
//!     unsafe_sp_chars: false,
//! };
//!
//! let c: char = char(opts);
//! println!("{}", c);
//! // ex: H
//! ```
//!
//! Or just use your own character set.
//!
//! ```
//! use keysmith::key;
//!
//! let charset = String::from("among_us");
//! let key: String = key::key_custom(32, charset);
//! println!("{}", key);
//!
//! // ex: unnusgnggauogoauusu__msguo_gmomg
//! ```
//!
//! Keysmith also lets you get a timestamp.
//! This is in seconds since the first second of 2022.
//! String is the default type for this, but you can also return an i64.
//! ```
//! use keysmith::timestamp;
//!
//! let stamp: String = timestamp::timestamp();
//! let stamp2: i64 = timestamp::timestamp_i64();
//!
//! println!("{}", stamp);
//! println!("{}", stamp2);
//! // ex: 25928414
//! ```
//!
//! Timestamps can also be in milliseconds.
//!
//! ```
//! use keysmith::timestamp;
//!
//! let stamp: String = timestamp::timestamp();
//! let stamp2: i64 = timestamp::timestamp_i64();
//!
//! println!("{}", stamp);
//! println!("{}", stamp2);
//! // ex: 25928414
//! ```
//!
//! Finally, you can get the current date or time.
//!
//! ```
//! use keysmith::date;
//!
//! let date: String = date::date();
//! let time: String = date::time();
//! let datetime: String = date::datetime();
//!
//! println!("{}", date); // ex: 2023-08-16
//! println!("{}", time); // ex: 19:29:10
//! println!("{}", datetime); // ex: 2023-08-16T19:29:10
//! ```
//!
//! Those were just the basics. For more functionality, see the modules below.

pub mod char;
pub mod date;
pub mod key;
mod test_utils;
pub mod timestamp;
pub mod uuid;
